packages/starlight/__e2e__/test-utils.ts

```diff
@@ -1,34 +1,74 @@
 import { fileURLToPath } from 'node:url';
 import { test as baseTest, type Page } from '@playwright/test';
-import { build, preview } from 'astro';
+import { build, dev, preview, type AstroInlineConfig } from 'astro';
 
 export { expect, type Locator } from '@playwright/test';
 
 process.env.ASTRO_TELEMETRY_DISABLED = 'true';
 
 // Setup a test environment that will build and start a preview server for a given fixture path and
 // provide a Starlight Playwright fixture accessible from within all tests.
-export async function testFactory(fixturePath: string) {
-	let previewServer: PreviewServer | undefined;
+export function testFactory(fixturePath: string) {
+	const fixturePathUrl = new URL(fixturePath, import.meta.url);
+	// Combining absolute paths with a `file:` base URL on Windows results
+	// in a URL with the drive letter as the protocol.
+	// In that case, use the URL as is instead of interpreting the `file:` protocol.
+	const root =
+		fixturePathUrl.protocol === 'file:'
+			? fileURLToPath(new URL(fixturePath, import.meta.url))
+			: fixturePathUrl.toString();
 
-	const test = baseTest.extend<{ starlight: StarlightPage }>({
-		starlight: async ({ page }, use) => {
-			if (!previewServer) {
-				throw new Error('Could not find a preview server to run tests against.');
-			}
+	async function makeServer(
+		options: {
+			mode?: 'build' | 'dev';
+			config?: AstroInlineConfig;
+		} = {}
+	): Promise<Server> {
+		const { mode, config } = options;
+		if (mode === 'dev') {
+			return await dev({ logLevel: 'error', root, ...config });
+		} else {
+			await build({ logLevel: 'error', root, ...config });
+			return await preview({ logLevel: 'error', root, ...config });
+		}
+	}
 
-			await use(new StarlightPage(previewServer, page));
-		},
-	});
+	const customServers: Server[] = [];
+	// Optimization for tests that don't customize any server options
+	// to not rebuild the fixture for each test.
+	const defaultServers: {
+		build?: Server;
+		dev?: Server;
+	} = {};
+
+	const test = baseTest.extend<{
+		makeServer: (config?: Parameters<typeof makeServer>[0]) => Promise<StarlightPage>;
+	}>({
+		makeServer: ({ page }, use) =>
+			use(async (params) => {
+				const { mode = 'build', config } = params ?? {};
+
+				if (config === undefined) {
+					const server = (defaultServers[mode] ??= await makeServer({ mode }));
+					return new StarlightPage(server, page);
+				}
 
-	test.beforeAll(async () => {
-		const root = fileURLToPath(new URL(fixturePath, import.meta.url));
-		await build({ logLevel: 'error', root });
-		previewServer = await preview({ logLevel: 'error', root });
+				const newServer = await makeServer(params);
+				customServers.push(newServer);
+
+				return new StarlightPage(newServer, page);
+			}),
```

Comment:
What do you think of exposing 2 helpers, e.g. `getServer` and `makeServer`. Most tests would just use `getServer` to get the default server and the few tests needing a second one would also use `makeServer`?

Reply:
What would be the difference between `getServer` and `makeServer` without any parameters?

Reply:
My idea was that `getServer` would never take arguments, and would always return the default prod server created. On the opposite, `makeServer` would require a mode to create a new one (I don't think we ever pass an inline config to `makeServer` so maybe this part could even be removed).